src/app/clusters/scenes-server/scenes-server.h

```diff
@@ -0,0 +1,102 @@
+#include <app/data-model/Nullable.h>
+#include <credentials/GroupDataProvider.h>
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace Scenes {
+
+class ScenesServer : public CommandHandlerInterface, public AttributeAccessInterface
+{
+public:
+    static ScenesServer & Instance();
+
+    CHIP_ERROR Init();
+
+    // CommandHanlerInterface
+    void InvokeCommand(HandlerContext & ctx) override;
+    CHIP_ERROR EnumerateAcceptedCommands(const ConcreteClusterPath & cluster, CommandIdCallback callback, void * context) override;
+    CHIP_ERROR EnumerateGeneratedCommands(const ConcreteClusterPath & cluster, CommandIdCallback callback, void * context) override;
+
+    // AttributeAccessInterface
+    CHIP_ERROR Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder) override;
+    CHIP_ERROR Write(const ConcreteDataAttributePath & aPath, AttributeValueDecoder & aDecoder) override;
+
+    // Callbacks
+    void OnGroupRemoved(const FabricIndex & aFabricIx, const EndpointId & aEndpointId, const GroupId & aGroupId);
+    void OnMakeInvalid();
+    void OnStoreCurrentScene(const FabricIndex & aFabricIx, const EndpointId & aEndpointId, const GroupId & aGroupId,
+                             const SceneId & aSceneId);
+    void OnRecallScene(const FabricIndex & aFabricIx, const EndpointId & aEndpointId, const GroupId & aGroupId,
+                       const SceneId & aSceneId);
+
+private:
+    ScenesServer() : CommandHandlerInterface(Optional<EndpointId>(), Id), AttributeAccessInterface(Optional<EndpointId>(), Id) {}
+    ~ScenesServer() {}
+
+    bool mIsInitialized = false;
+
+    //  Command handlers
+    void HandleAddScene(HandlerContext & ctx, const Commands::AddScene::DecodableType & req);
+    void HandleViewScene(HandlerContext & ctx, const Commands::ViewScene::DecodableType & req);
+    void HandleRemoveScene(HandlerContext & ctx, const Commands::RemoveScene::DecodableType & req);
+    void HandleRemoveAllScenes(HandlerContext & ctx, const Commands::RemoveAllScenes::DecodableType & req);
+    void HandleStoreScene(HandlerContext & ctx, const Commands::StoreScene::DecodableType & req);
+    void HandleRecallScene(HandlerContext & ctx, const Commands::RecallScene::DecodableType & req);
+    void HandleGetSceneMembership(HandlerContext & ctx, const Commands::GetSceneMembership::DecodableType & req);
+    void HandleEnhancedAddScene(HandlerContext & ctx, const Commands::EnhancedAddScene::DecodableType & req);
+    void HandleEnhancedViewScene(HandlerContext & ctx, const Commands::EnhancedViewScene::DecodableType & req);
+    void HandleCopyScene(HandlerContext & ctx, const Commands::CopyScene::DecodableType & req);
+
+    ConcreteCommandPath mPath = ConcreteCommandPath(0, 0, 0);
+
+    // Feature Flag
+    BitFlags<ScenesFeature> mFeatureFlags;
+
+    // Attributes
+    uint8_t mSceneCount   = 0;
+    SceneId mCurrentScene = 0;
+    GroupId mCurrentGroup = 0;
+    bool mSceneValid      = false;
```

Comment:
These three are per-endpoint, no?

Reply:
Will get fixed in next commit with the endpoint scope rework.